src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -4040,26 +4301,356 @@ ColocationGroupCreateCommandList(void)
+	 */
+	dependencies = FilterObjectAddressListByPredicate(dependencies,
+													  &SupportedDependencyByCitus);
+
+	dependencies = OrderObjectAddressListInDependencyOrder(dependencies);
+
+	MemoryContext commandsContext = AllocSetContextCreate(CurrentMemoryContext,
+														  "dependency commands context",
+														  ALLOCSET_DEFAULT_SIZES);
+	MemoryContextSwitchTo(commandsContext);
+
+	ObjectAddress *dependency = NULL;
+	foreach_ptr(dependency, dependencies)
 	{
-		return NIL;
+		if (IsAnyObjectAddressOwnedByExtension(list_make1(dependency), NULL))
+		{
+			/*
+			 * we expect extension-owned objects to be created as a result
+			 * of the extension being created.
+			 */
+			continue;
+		}
+
+		if (!context->commandAccumulationForTest)
+		{
+			MemoryContextReset(commandsContext);
+		}
+
+		/* dependency creation commands */
+		List *ddlCommands = GetAllDependencyCreateDDLCommands(list_make1(dependency));
+
+		if (ddlCommands == NIL)
+		{
+			continue;
+		}
+
+		ddlCommands = lcons(DISABLE_DDL_PROPAGATION,
+							ddlCommands);
+		ddlCommands = lappend(ddlCommands,
+							  ENABLE_DDL_PROPAGATION);
+
+		SendCommandListToActivatedNodesViaMetadataContext(context, ddlCommands);
+	}
+
+	MemoryContextSwitchTo(oldContext);
+
+	if (!context->commandAccumulationForTest)
+	{
+		MemoryContextDelete(commandsContext);
+	}
+	ResetMetadataSyncMemoryContext(context);
+}
+
+
+/*
+ * SendMetadataCommandsViaMetadataContext sends commands related to pg_dist_shard and,
+ * pg_dist_shard_placement entries to workers with transactional or nontransactional mode
+ * according to the flag `useCoordinatedTran` inside metadataSyncContext.
+ */
+void
+SendMetadataCommandsViaMetadataContext(MetadataSyncContext *context)
+{
+	ScanKeyData scanKey[1];
+	int scanKeyCount = 0;
+
+	Relation relation = table_open(DistPartitionRelationId(), AccessShareLock);
+
+	TupleDesc tupleDesc = RelationGetDescr(relation);
+	SysScanDesc scanDesc = systable_beginscan(relation,
+											  InvalidOid, false,
+											  NULL, scanKeyCount, scanKey);
+
+	MemoryContext oldContext = MemoryContextSwitchTo(context->context);
+
+	HeapTuple nextTuple = NULL;
+	while (true)
+	{
+		ResetMetadataSyncMemoryContext(context);
+
+		MemoryContextSwitchTo(oldContext);
+		nextTuple = systable_getnext(scanDesc);
+		if (!HeapTupleIsValid(nextTuple))
+		{
+			MemoryContextSwitchTo(context->context);
+			break;
+		}
+		MemoryContextSwitchTo(context->context);
+
+		/*
+		 * create Citus table metadata commands (pg_dist_shard, pg_dist_shard_placement,
+		 * pg_dist_partition). Only Citus tables have shard metadata.
+		 */
+		Oid relationId = FetchRelationIdFromPgPartitionHeapTuple(nextTuple, tupleDesc);
+
+		if (!ShouldSyncTableMetadata(relationId))
+		{
+			continue;
+		}
+
+		List *tableMetadataCreateCommandList =
+			CitusTableMetadataCreateCommandList(relationId);
+
+		tableMetadataCreateCommandList = lcons(DISABLE_DDL_PROPAGATION,
```

Comment:
why do we require `ENABLEE|DISABLE_DDL_PROPAGATION`  for sending pg_dist_shard/placement?